File: An2/MAP/MAP-Java/MAP-Lab5/src/Repository/Stack.java
/*
 * To change this template, choose Tools | Templates
 * and open the template in the editor.
 */
package Repository;

import Controller.StudentException;
import Domain.Student;

/**
 *
 * @author mihai
 */
public class Stack {
    private int size;
    private int capacity;
    private Student[] students;
    
    public Stack(int sCapacity){
        this.size = -1;
        this.capacity = sCapacity;
        this.students = new Student[sCapacity];
    }
    

    
    public void push(Student s){
        size++;
        if (size == capacity){
            this.increaseCapacity();
        }
        students[size] = s;
    }
    
    public Student pop() throws StudentException{
        if (this.isEmpty())
            throw new StudentException("Cannot pop elem from empty stack!");
        else {
            Student student = this.students[size];
            this.students[size] = null;
            this.size--;
            return student;
        }
    }
    
    private void increaseCapacity() {
        capacity = capacity*2;
    }
   
    public int getStackSize(){
        return this.size;
    }
    
    public int getStackCapacity(){
        return this.capacity;
    }
    
    private boolean isEmpty(){
        return this.size < 0;
    }

}

File: An2/MAP/MAP-C#/MAP-Lab5/MAP-Lab5/Repository/Stack.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

using MAP_Lab5.Domain;
using MAP_Lab5.Controller;

namespace MAP_Lab5.Repository
{
    class Stack
    {
        private int size;
        private int capacity;
        private Student[] students;

        public Stack(int capacity)
        {
            /*
            * pre cond: int sCapacity - maximum number of students to be added
            * post cond: object Stack
            * 
            * creates Stack, allocates memory
            */

            students = new Student[capacity];
            size = -1;
            this.capacity = capacity;
        }

        public void push(Student s)
        {
            /*
            * pre cond: s of type Student
            * post cond: -
            * 
            * adds object on top of stack; if maximum capacity is reached,
            * it doubles it
            */


            size++;
            if (size == capacity)
                increaseCapacity();
            students[size] = s;
        }

        public Student pop()
        {
            /*
            * pre cond: -
            * post cond: the last added Student
            * 
            * removes the object on top of the stack and returns it
            */
            if (isEmpty())
                throw new StudentException("Cannot pop from empty stack!");
            else
            {
                Student student = students[size];
                students[size] = null;
                size--;
                return student;
            }
        }

        private void increaseCapacity()
        {
            /*
            * pre cond: -
            * post cond: the capacity is doubled
            * 
            * doubles the current capacity of the stack
            */

            capacity *= 2;
        }

        private bool isEmpty()
        {
            if (this.size < 0)
                return true;
            return false;
        }

        public int getStackSize()
        {
            return this.size;
        }

        public int getStackCapacity()
        {
            return this.capacity;
        }

    }
}
